internal/expr/bindinputs.go

```diff
@@ -52,31 +40,86 @@ func (te *TypedExpr) BindInputs(args ...any) (pq *PrimedQuery, err error) {
 			return nil, fmt.Errorf("type %q provided more than once", t.Name())
 		}
 		typeToValue[t] = v
-		if !inQuery[t] {
-			// Check if we have a type with the same name from a different package.
-			for _, input := range te.inputs {
-				if t.Name() == input.ArgType().Name() {
-					return nil, fmt.Errorf("parameter with type %q missing, have type with same name: %q", input.ArgType().String(), t.String())
-				}
-			}
-			return nil, fmt.Errorf("%s not referenced in query", t.Name())
-		}
 	}
 
-	// Query parameters.
+	// Generate SQL and query parameters.
 	var params []any
+	var outputs []typeinfo.Output
+	// argTypeUsed is used to check that all the query parameters are
+	// referenced in the query.
+	argTypeUsed := map[reflect.Type]bool{}
 	inputCount := 0
-	for _, input := range te.inputs {
-		vals, err := input.LocateParams(typeToValue)
-		if err != nil {
-			return nil, err
+	outputCount := 0
+	sqlStr := bytes.Buffer{}
+	for _, te := range *tbe {
+		switch te := te.(type) {
+		case *typedInputExpr:
+			vals, err := te.input.LocateParams(typeToValue)
+			if err != nil {
+				return nil, err
+			}
+			argTypeUsed[te.input.ArgType()] = true
+			for _, val := range vals {
+				namedInput := sql.Named("sqlair_"+strconv.Itoa(inputCount), val.Interface())
+				params = append(params, namedInput)
+				sqlStr.WriteString("@sqlair_" + strconv.Itoa(inputCount))
+				inputCount++
+			}
+		case *typedOutputExpr:
+			for i, oc := range te.outputColumns {
+				sqlStr.WriteString(oc.sql(outputCount))
+				if i != len(te.outputColumns)-1 {
+					sqlStr.WriteString(", ")
+				}
+				outputCount++
+				outputs = append(outputs, oc.output)
+			}
+		case *bypass:
+			sqlStr.WriteString(te.chunk)
+		default:
+			return nil, fmt.Errorf("invalid expression type %t", te)
```

Comment:
My mistake. This should be `%T`.

Reply:
Fixed, also added the internal error prefix.